documentos/canaima-desarrollador.1
.\" Man page generated from reStructeredText.
.
.TH CANAIMA-DESARROLLADOR 1 "2011-01-22" "1.0+0" "Empaquetamiento"
.SH NAME
CANAIMA-DESARROLLADOR \- Un conjunto de herramientas para la eliminación de las barreras tecnológicas
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH MODO DE USO
.sp
.nf
.ft C
canaima\-desarrollador [AYUDANTE] [PARÁMETRO\-1] [PARÁMETRO\-2] ... [PARÁMETRO\-N] [\-\-ayuda]
.ft P
.fi
.SH DESCRIPCIÓN
.sp
Canaima Desarrollador (C\-D) es un compendio de herramientas y ayudantes que facilitan el proceso de desarrollo de software para Canaima GNU/Linux. Está diseñado para \fBfacilitar el trabajo\fP a aquellas personas que participan en dicho proceso con regularidad, como también para \fBiniciar a los que deseen aprender\fP de una manera rápida y práctica.
.sp
C\-D puede ayudarte a:
.INDENT 0.0
.IP \(bu 2
.
Agilizar los procesos para la creación de paquetes binarios canaima a partir de paquetes fuentes correctamente estructurados.
.IP \(bu 2
.
Automatización personalizada de la creación de Paquetes Fuentes acordes a las Políticas de Canaima GNU/Linux.
.IP \(bu 2
.
Creación de un depósito personal, por usuario, donde se guardan automáticamente y en carpetas separadas los siguientes tipos de archivo:
.INDENT 2.0
.IP \(bu 2
.
Proyectos en proceso de empaquetamiento
.IP \(bu 2
.
Paquetes Binarios (*.deb)
.IP \(bu 2
.
Paquetes Fuente (*.tar.gz, *.dsc, *.changes, *.diff)
.IP \(bu 2
.
Registros provenientes de la creación de paquetes binarios (*.build)
.UNINDENT
.IP \(bu 2
.
Versionamiento asistido (basado en git) en los proyectos, brindando herramientas para realizar las siguientes operaciones, con un alto nivel de automatización y detección de posibles errores:
.INDENT 2.0
.IP \(bu 2
.
git clone
.IP \(bu 2
.
git commit
.IP \(bu 2
.
git push
.IP \(bu 2
.
git pull
.UNINDENT
.IP \(bu 2
.
Ejecución de tareas en masa (empaquetar, hacer pull, push, commit, entre otros), para agilizar procesos repetitivos.
.UNINDENT
.SH AYUDANTES DE CANAIMA DESARROLLADOR
.INDENT 0.0
.IP \(bu 2
.
Crear Proyecto / Debianizar
.IP \(bu 2
.
Crear Fuente
.IP \(bu 2
.
Empaquetar
.IP \(bu 2
.
Descargar
.IP \(bu 2
.
Registrar
.IP \(bu 2
.
Enviar
.IP \(bu 2
.
Actualizar
.IP \(bu 2
.
Descargar Todo
.IP \(bu 2
.
Registrar Todo
.IP \(bu 2
.
Enviar Todo
.IP \(bu 2
.
Actualizar Todo
.IP \(bu 2
.
Empaquetar Varios
.IP \(bu 2
.
Empaquetar Todo
.IP \(bu 2
.
Listar Remotos
.IP \(bu 2
.
Listar Locales
.UNINDENT
.SH CREAR PROYECTO / DEBIANIZAR
.sp
Crea un proyecto de empaquetamiento desde cero o debianiza uno existente.
.SS USO
.sp
.nf
.ft C
canaima\-desarrollador crear\-proyecto|debianizar \-\-nombre="" \-\-version="" \-\-destino="" \-\-licencia="" [\-\-ayuda]
.ft P
.fi
.SS PARÁMETROS
.INDENT 0.0
.TP
.B \fB\-\-nombre\fP
.sp
Un nombre para tu proyecto, que puede contener letras, números, puntos y guiones. Cualquier otro caracter no está permitido.
.TP
.B \fB\-\-version\fP
.sp
La versión inicial de tu proyecto. Se permiten números, guiones, puntos, letras o dashes (~).
.TP
.B \fB\-\-destino\fP
.sp
Especifica si es un proyecto de empaquetamiento para Canaima GNU/Linux o si es un proyecto personal. Las opciones disponibles son "canaima" y "personal".
.TP
.B \fB\-\-licencia\fP
.sp
Especifica el tipo de licencia bajo el cuál distribuirás tu trabajo. Las licencias soportadas son: apache, artistic, bsd, gpl, gpl2, gpl3, lgpl, lgpl2 y lgpl3.
.TP
.B \fB\-\-ayuda\fP
.sp
Muestra la documentación para el ayudante.
.UNINDENT
.sp
Si estás debianizando un proyecto existente, lo que ingreses en \-\-nombre="proyecto" y \-\-version="X.Y+Z" se utilizará para determinar cuál es el nombre de la carpeta a debianizar dentro del directorio del desarrollador, suponiendo que tiene el nombre proyecto\-X.Y+Z. Si no se llama así, habrá un error.
.SH CREAR FUENTE
.sp
Crea un paquete fuente a partir de un proyecto de empaquetamiento existente. El resultado es guardado en el depósito de fuentes.
.SS USO
.sp
.nf
.ft C
canaima\-desarrollador crear\-fuente \-\-directorio="" [\-\-ayuda]
.ft P
.fi
.SS PARÁMETROS
.INDENT 0.0
.TP
.B \fB\-\-directorio\fP
.sp
Nombre del directorio dentro de la carpeta del desarrollador donde se encuentra el proyecto. El directorio debe contener un proyecto debianizado.
.TP
.B \fB\-\-ayuda\fP
.sp
Muestra la documentación para el ayudante.
.UNINDENT
.SH EMPAQUETAR
.sp
Éste ayudante te permite empaquetar un proyecto de forma automatizada, siguiendo la metodología git\-buildpackage, que se centra en el siguiente diagrama:
.sp
\fBCOMMIT > REFLEJAR CAMBIOS EN EL CHANGELOG > COMMIT > CREAR PAQUETE FUENTE > PUSH > GIT\-BUILDPACKAGE\fP
.SS USO
.sp
.nf
.ft C
canaima\-desarrollador empaquetar \-\-directorio="" \-\-mensaje="" \-\-procesadores="" [\-\-ayuda]
.ft P
.fi
.SS PARÁMETROS
.INDENT 0.0
.TP
.B \fB\-\-directorio\fP
.sp
Nombre de la carpeta dentro del directorio del desarrollador donde se encuentra el proyecto a empaquetar.
.TP
.B \fB\-\-mensaje\fP
.sp
Mensaje representativo de los cambios para el primer commit. El segundo commit es sólo para el changelog. Colocando la palabra "auto" o dejando el campo vacío, se autogenera el mensaje.
.TP
.B \fB\-\-procesadores\fP
.sp
Número de procesadores con que cuenta tu computadora para optimizar el proceso de empaquetamiento.
.TP
.B \fB\-\-ayuda\fP
.sp
Muestra la documentación para el ayudante.
.UNINDENT
.SH DESCARGAR
.sp
Éste ayudante te permite copiar a tu disco duro un proyecto que se encuentre en el repositorio remoto para que puedas modificarlo según consideres. Utiliza git clone para realizar tal operación. Éste ayudante se encarga además de realizar las siguientes operaciones por ti:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
.
Verifica e informa sobre el éxito de la descarga.
.UNINDENT
.UNINDENT
.UNINDENT
.SS USO
.sp
.nf
.ft C
canaima\-desarrollador descargar \-\-proyecto="" [\-\-ayuda]
.ft P
.fi
.SS PARÁMETROS
.INDENT 0.0
.TP
.B \fB\-\-proyecto\fP
.sp
Nombre del proyecto (en caso de que éste se encuentre en el repositorio de Canaima GNU/Linux) o la dirección git pública del proyecto.
.TP
.B \fB\-\-ayuda\fP
.sp
Muestra la documentación para el ayudante.
.UNINDENT
.SH REGISTRAR
.sp
Éste ayudante te permite registar (o hacer commit de) los cambios hechos en un proyecto mediante el versionamiento basado en git. Utiliza git commit para lograr éste propósito. Éste ayudante se encarga además de realizar las siguientes operaciones por ti:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
.
Verifica la existencia de la rama git "upstream". En caso de no encontrarla, la crea.
.IP \(bu 2
.
Verifica la existencia de la rama git "master". En caso de no encontrarla, la crea.
.IP \(bu 2
.
Verifica la existencia de todos los elementos necesarios para ejecutar la acción git commit (carpetas, variables de entorno, etc..). En caso de encontrar algún error, aborta e informa.
.IP \(bu 2
.
Autogenera el mensaje de commit, si se le instruye.
.IP \(bu 2
.
Hace git checkout a la rama master, si nos encontramos en una rama diferente a la hora de hace commit.
.IP \(bu 2
.
Hace un git merge de la rama master a la upstream, inmediatamente depués del commit.
.UNINDENT
.UNINDENT
.UNINDENT
.SS USO
.sp
.nf
.ft C
canaima\-desarrollador registrar \-\-directorio="" \-\-mensaje=="" [\-\-ayuda]
.ft P
.fi
.SS PARÁMETROS
.INDENT 0.0
.TP
.B \fB\-\-directorio\fP
.sp
Nombre de la carpeta dentro del directorio del desarrollador a la que se quiere hacer commit.
.TP
.B \fB\-\-mensaje\fP
.sp
Mensaje representativo de los cambios para el commit. Colocando la palabra "auto" o dejando el campo vacío, se autogenera el mensaje.
.TP
.B \fB\-\-ayuda\fP
.sp
Muestra la documentación para el ayudante.
.UNINDENT
.SH ENVIAR
.sp
Éste ayudante te permite enviar los cambios realizados al repositorio remoto especificado en las configuraciones personales, mediante el uso de la acción git push. Éste ayudante se encarga además de realizar las siguientes operaciones por ti:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
.
Verifica la existencia de la rama git "upstream". En caso de no encontrarla, la crea.
.IP \(bu 2
.
Verifica la existencia de la rama git "master". En caso de no encontrarla, la crea.
.IP \(bu 2
.
Verifica la existencia de todos los elementos necesarios para ejecutar la acción git push (carpetas, variables de entorno, etc..). En caso de encontrar algún error, aborta e informa.
.IP \(bu 2
.
Configura el repositorio remoto para el proyecto, de acuerdo a los parámetros establecidos en ~/.config/canaima\-desarrollador/usuario.conf
.UNINDENT
.UNINDENT
.UNINDENT
.SS USO
.sp
.nf
.ft C
canaima\-desarrollador enviar \-\-directorio="" [\-\-ayuda]
.ft P
.fi
.SS PARÁMETROS
.INDENT 0.0
.TP
.B \fB\-\-directorio\fP
.sp
Nombre de la carpeta dentro del directorio del desarrollador a la que se quiere hacer push.
.TP
.B \fB\-\-ayuda\fP
.sp
Muestra la documentación para el ayudante.
.UNINDENT
.SH ACTUALIZAR
.sp
Éste ayudante te permite actualizar el código fuente de un determinado proyecto, mediante la ejecución de "git pull" en la carpeta del proyecto. Éste ayudante se encarga además de realizar las siguientes operaciones por ti:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
.
Verifica la existencia de la rama git "upstream". En caso de no encontrarla, la crea.
.IP \(bu 2
.
Verifica la existencia de la rama git "master". En caso de no encontrarla, la crea.
.IP \(bu 2
.
Verifica la existencia de todos los elementos necesarios para ejecutar la acción git pull (carpetas, variables de entorno, etc..). En caso de encontrar algún error, aborta e informa.
.IP \(bu 2
.
Configura el repositorio remoto para el proyecto, de acuerdo a los parámetros establecidos en ~/.config/canaima\-desarrollador/usuario.conf
.UNINDENT
.UNINDENT
.UNINDENT
.SS USO
.sp
.nf
.ft C
canaima\-desarrollador actualizar \-\-directorio="" [\-\-ayuda]
.ft P
.fi
.SS PARÁMETROS
.INDENT 0.0
.TP
.B \fB\-\-directorio\fP
.sp
Nombre de la carpeta dentro del directorio del desarrollador a la que se quiere hacer git pull.
.TP
.B \fB\-\-ayuda\fP
.sp
Muestra la documentación para el ayudante.
.UNINDENT
.SH DESCARGAR TODO
.sp
Éste ayudante te permite copiar a tu disco duro todos los proyectos de Canaima GNU/Linux que se encuentren en el repositorio remoto oficial. Utiliza git clone para realizar tal operación.
.SS USO
.sp
.nf
.ft C
canaima\-desarrollador descargar\-todo [\-\-ayuda]
.ft P
.fi
.SS PARÁMETROS
.INDENT 0.0
.TP
.B \fB\-\-ayuda\fP
.sp
Muestra la documentación para el ayudante.
.UNINDENT
.SH REGISTRAR TODO
.sp

debian/changelog
canaima-desarrollador (1.1+39) testing; urgency=low

  * [455b67d] Removiendo dependencia a un navegador y la creación de la
    entrada de menú
  * [ad985a9] Removiendo dependencia a un navegador y la creación de la
    entrada de menú
  * [b362440] Removiendo git-tag del empaquetador
  * [f4e2017] Removiendo git-tag del empaquetador
  * [59a296f] Removiendo git-tag del empaquetador
  * [0fe9526] Removiendo git-tag del empaquetador

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Wed, 27 Jul 2011 08:52:34 -0430

canaima-desarrollador (1.1+38) testing; urgency=low

  * [c13418f] Cambaindo la dependencia de construcción de texlive-latex-
    extra a texlive, es más completo.

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Wed, 18 May 2011 01:04:53 -0430

canaima-desarrollador (1.1+37) testing; urgency=low

  * [1fcf907] Agregando dependencia a lintian, para asegurar la revisión
    de calidad.
  * [56d290e] Agregando dependencia a lintian, para asegurar la revisión
    de calidad.
  * [8b856be] Agregando dependencia a lintian, para asegurar la revisión
    de calidad.

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Wed, 18 May 2011 00:56:44 -0430

canaima-desarrollador (1.1+36) testing; urgency=low

  * [a23f02f] Moviendo la ubicación del .desktop a GTK;Development;IDE.

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Fri, 15 Apr 2011 01:15:49 -0430

canaima-desarrollador (1.1+35) testing; urgency=low

  * [3d4451b] Agregando dependencias a canaima-curiara o www-browser.

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Sun, 27 Mar 2011 12:33:17 -0430

canaima-desarrollador (1.1+34) testing; urgency=low

  * [88af140] Agregando dependencias fakeroot y build-essential.

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Sat, 26 Mar 2011 11:51:06 -0430

canaima-desarrollador (1.1+33) testing; urgency=low

  * [f013915] Agregando Carlos Escobar a Uploaders.

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Tue, 22 Mar 2011 14:38:26 -0430

canaima-desarrollador (1.1+32) testing; urgency=low

  * [48f4951] Arreglando forma en que se copia el Manual del
    Desarrolldor en el Escritorio.

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Tue, 22 Mar 2011 14:22:05 -0430

canaima-desarrollador (1.1+31) testing; urgency=low

  * [658a843] [ canaima-desarrollador ] Los siguientes archivos han sido
    modificados/añadidos: funciones-desarrollador.sh

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Fri, 18 Mar 2011 00:31:53 -0430

canaima-desarrollador (1.1+30) testing; urgency=low

  * [c19033f] [ canaima-desarrollador ] Los siguientes archivos han sido
    modificados/añadidos: funciones-desarrollador.sh

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Fri, 18 Mar 2011 00:29:36 -0430

canaima-desarrollador (1.1+29) testing; urgency=low

  * [05731d3] [ canaima-desarrollador ] Los siguientes archivos han sido
    modificados/añadidos: funciones-desarrollador.sh

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Thu, 17 Mar 2011 23:48:03 -0430

canaima-desarrollador (1.1+28) testing; urgency=low

  * [d211c7a] [ canaima-desarrollador ] Los siguientes archivos han sido
    modificados/añadidos: funciones-desarrollador.sh

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Thu, 17 Mar 2011 16:59:46 -0430

canaima-desarrollador (1.1+27) testing; urgency=low

  * [4513efc] [ canaima-desarrollador ] Los siguientes archivos han sido
    modificados/añadidos: control funciones-desarrollador.sh

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Thu, 17 Mar 2011 16:49:26 -0430

canaima-desarrollador (1.1+26) testing; urgency=low

  * [0e16eeb] [ canaima-desarrollador ] Los siguientes archivos han sido
    modificados/añadidos: postinst hola1.deb hola2.deb hola3.deb
    funciones-desarrollador.sh

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Thu, 17 Mar 2011 16:12:13 -0430

canaima-desarrollador (1.1+25) testing; urgency=low

  * [37c0134] Mejorando Makefile y corrigiendo bug del cambio de
    directorio mientras se empaqueta
  * [122451d] Corrigiendo error en el Makefile.
  * [6a59e1f] Borrando residuo de compilación.
  * [a57f3b1] Mejorando la forma en que se firman los paquetes.
  * [0efc847] Agregando opción "--no-enviar" cuando se empaqueta

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Mon, 14 Mar 2011 17:10:24 -0430

canaima-desarrollador (1.1+24) testing; urgency=low

  * [06befe6] Mejorando documentación de ayuda

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Mon, 14 Mar 2011 14:49:19 -0430

canaima-desarrollador (1.1+23) testing; urgency=low

  * [da36c13] Removiendo residuos y haciendo más específica la forma en
    que se determina que sólo se modificó el debian/changelog
  * [b954ce8] Mejorando la forma en que se envían las tags

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Mon, 14 Mar 2011 14:32:59 -0430

canaima-desarrollador (1.1+22) testing; urgency=low

  * [49f236d] Especificando mensaje de Nueva versión cuando se hace el
    commit de sólo debian/changelog

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Mon, 14 Mar 2011 13:51:22 -0430

canaima-desarrollador (1.1+21) testing; urgency=low

  * [381f83c] Corrigiendo bug con la ejecución de los parámetros de
    ayuda.

 -- Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>  Mon, 14 Mar 2011 13:22:21 -0430


scripts/canaima-desarrollador.sh
#!/bin/bash -e
#
# ==============================================================================
# PAQUETE: canaima-desarrollador
# ARCHIVO: canaima-desarrollador.sh
# DESCRIPCIÓN: Script de bash principal del paquete canaima-desarrollador
# COPYRIGHT:
#  (C) 2010 Luis Alejandro Martínez Faneyth <martinez.faneyth@gmail.com>
#  (C) 2010 Diego Alberto Aguilera Zambrano <daguilera85@gmail.com>
#  (C) 2010 Carlos Alejandro Guerrero Mora <guerrerocarlos@gmail.com>
#  (C) 2010 Francisco Javier Vásquez Guerrero <franjvasquezg@gmail.com>
# LICENCIA: GPL3
# ==============================================================================
#
# Este programa es software libre. Puede redistribuirlo y/o modificarlo bajo los
# términos de la Licencia Pública General de GNU (versión 3).

VARIABLES="/usr/share/canaima-desarrollador/variables.conf"

# Inicializando variables
. ${VARIABLES}

# Cargando configuración
. ${CONF}

# Cargando funciones
. ${FUNCIONES}

# Función para chequear que ciertas condiciones se cumplan para el
# correcto funcionamiento de canaima-desarrollador
# Ver /usr/share/canaima-desarrollador/scripts/funciones-desarrollador.sh
CHECK

# Función para cargar los datos del desarrollador especificados en
# ${CONF} (nombre, correo, etc.)
# Ver /usr/share/canaima-desarrollador/scripts/funciones-desarrollador.sh
DEV-DATA

# Capturamos los parámetros 
PARAMETROS=${@}

# Si hay parámetros, y no existen parámetros de ayuda ...
if [ -n "${PARAMETROS}" ] && [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) == 0 ] && [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) == 0 ]; then

# Removemos el ayudante
PARAMETROS=${PARAMETROS#${1}}

# Damos formato a los parámetros para que sean seleccionados correctamente por el ciclo for
PARAMETROS=$( echo ${PARAMETROS} | sed 's/=/="/g' )
PARAMETROS=$( echo ${PARAMETROS} | sed 's/ --/"#####--/g' )
PARAMETROS=$( echo ${PARAMETROS} | sed 's/ /_____/g' )
PARAMETROS=$( echo ${PARAMETROS} | sed 's/#####/ /g' )
[ -n "${PARAMETROS}" ] && PARAMETROS="${PARAMETROS}\""

# Parseamos cada parámetro y lo convertimos en variable
for ARGUMENTO in ${PARAMETROS}; do
ARG_VARIABLE=${ARGUMENTO#--}
ARG_VARIABLE=${ARG_VARIABLE%=*}
ARG_VALOR=${ARGUMENTO#--${ARG_VARIABLE}=}
ARG_VALOR=$( echo ${ARG_VALOR} | sed 's/_____/ /g' )
ARG_VARIABLE=$( echo ${ARG_VARIABLE} | tr '[:lower:]' '[:upper:]' )
ARG_VARIABLE=$( echo ${ARG_VARIABLE} | tr '-' '_' )
eval "${ARG_VARIABLE}=${ARG_VALOR}"
done

fi

# Case encargado de interpretar los parámetros introducidos a
# canaima-desarrollador y ejecutar la función correspondiente
case ${1} in

#------ AYUDANTES CREADORES --------------------------------------------------------------------------------------#
#=================================================================================================================#

crear-proyecto|debianizar)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/crear-proyecto"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-nombre" ) == 0 ] && [ $( echo ${VERIFICAR} | grep -c "\-\-version" ) == 0 ] && [ $( echo ${VERIFICAR} | grep -c "\-\-destino" ) == 0 ] && [ $( echo ${VERIFICAR} | grep -c "\-\-licencia" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

# Guardemos los parámetros en variables para usarlos después
opcion=${1}
nombre=${NOMBRE}
version=${VERSION}
destino=${DESTINO}
licencia=${LICENCIA}
CREAR-PROYECTO
fi
;;

crear-fuente)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/crear-fuente"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-directorio" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

# Guardando directorio en variable para utilizarlo después
directorio=${DIRECTORIO}
CREAR-FUENTE
fi
;;

empaquetar)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/empaquetar"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-directorio" ) == 0 ] && [ $( echo ${VERIFICAR} | grep -c "\-\-mensaje" ) == 0 ] && [ $( echo ${VERIFICAR} | grep -c "\-\-procesadores" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

# Guardamos los parámetros en variables para usarlas después
directorio=${DIRECTORIO}
mensaje=${MENSAJE}
procesadores=${PROCESADORES}
EMPAQUETAR
fi
;;

#------ AYUDANTES GIT --------------------------------------------------------------------------------------------#
#=================================================================================================================#
descargar|clonar|clone)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/descargar"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-proyecto" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

# Guardemos el segundo argumento en la varible "proyecto"
proyecto=${PROYECTO}
# Ejecutemos la función correspondiente
DESCARGAR
fi
;;

registrar|commit)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/registrar"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-directorio" ) == 0 ] && [ $( echo ${VERIFICAR} | grep -c "\-\-mensaje" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

# Guardemos los parámetros en variables para usarlos después
directorio=${DIRECTORIO}
mensaje=${MENSAJE}
REGISTRAR
fi
;;

enviar|push)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/enviar"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-directorio" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

# Guardando directorio en variable para utilizarlo después
directorio=${DIRECTORIO}
ENVIAR
fi
;;

actualizar|pull)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/actualizar"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-directorio" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

# Guardando directorio en variable para utilizarlo después
directorio=${DIRECTORIO}
ACTUALIZAR
fi
;;

#------ AYUDANTES MASIVOS ----------------------------------------------------------------------------------------#
#=================================================================================================================#

descargar-todo|clonar-todo)
if [ "${2}" == "--ayuda" ] || [ $"{2}" == "--help" ]; then
cat "${DIR_AYUDA}/descargar-todo"
else
[ -n "${2}" ] && ERROR "No conozco la opción '${2}', revisa la documentación." && exit 1
DESCARGAR-TODO
fi
;;

registrar-todo|commit-todo)
if [ "${2}" == "--ayuda" ] || [ $"{2}" == "--help" ]; then
cat "${DIR_AYUDA}/registrar-todo"
else
[ -n "${2}" ] && ERROR "No conozco la opción '${2}', revisa la documentación." && exit 1
mensaje="auto"
REGISTRAR-TODO
fi
;;

enviar-todo|push-todo)
if [ "${2}" == "--ayuda" ] || [ $"{2}" == "--help" ]; then
cat "${DIR_AYUDA}/enviar-todo"
else
[ -n "${2}" ] && ERROR "No conozco la opción '${2}', revisa la documentación." && exit 1
ENVIAR-TODO
fi
;;

actualizar-todo|pull-todo)
if [ "${2}" == "--ayuda" ] || [ $"{2}" == "--help" ]; then
cat "${DIR_AYUDA}/actualizar-todo"
else
[ -n "${2}" ] && ERROR "No conozco la opción '${2}', revisa la documentación." && exit 1
ACTUALIZAR-TODO
fi
;;

empaquetar-varios)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/empaquetar-varios"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-para-empaquetar" ) == 0 ] && [ $( echo ${VERIFICAR} | grep -c "\-\-procesadores" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

# Guardamos los parámetros en variables para usarlas después
para_empaquetar=${PARA_EMPAQUETAR}
mensaje="auto"
procesadores=${PROCESADORES}
EMPAQUETAR-VARIOS
fi
;;

empaquetar-todo)
if [ -z "${PARAMETROS}" ] || [ $( echo ${PARAMETROS} | grep -c "\-\-ayuda" ) != 0 ] || [ $( echo ${PARAMETROS} | grep -c "\-\-help" ) != 0 ]; then
cat "${DIR_AYUDA}/empaquetar-todo"
else

for VERIFICAR in ${PARAMETROS}; do
[ $( echo ${VERIFICAR} | grep -c "\-\-procesadores" ) == 0 ] && ERROR "No conozco la opción '${VERIFICAR}', revisa la documentación." && exit 1
done

mensaje="auto"
procesadores=${PROCESADORES}
EMPAQUETAR-TODO
fi
;;


#------ AYUDANTES INFORMATIVOS -----------------------------------------------------------------------------------------#
#=======================================================================================================================#

listar-remotos)
if [ "${2}" == "--ayuda" ] || [ $"{2}" == "--help" ]; then
cat "${DIR_AYUDA}/listar-remotos"
else
[ -n "${2}" ] && ERROR "No conozco la opción '${2}', revisa la documentación." && exit 1
LISTAR-REMOTOS
fi
;;

listar-locales)
if [ "${2}" == "--ayuda" ] || [ $"{2}" == "--help" ]; then
cat "${DIR_AYUDA}/listar-locales"
else
[ -n "${2}" ] && ERROR "No conozco la opción '${2}', revisa la documentación." && exit 1
LISTAR-LOCALES
fi
;;

--ayuda|--help|'')
[ -n "${2}" ] && ERROR "No conozco la opción '${2}', revisa la documentación." && exit 1
# Imprimiendo la ayuda
cat "${DIR_AYUDA}/canaima-desarrollador"
;;

debug)
[ -n "${2}" ] && ERROR "No conozco la opción '${2}', revisa la documentación." && exit 1
LISTA_DATOS="DEV_NAME DEV_MAIL DEV_GPG DEV_DIR DEPOSITO_SOURCES DEPOSITO_LOGS DEPOSITO_DEBS REPO REPO_USER REPO_DIR CONF VARIABLES PLANTILLAS FUNCIONES NAVEGADOR"
for VARIABLES_CONF in ${LISTA_DATOS}; do
eval "VARIABLES_CONF_EVAL=\$${VARIABLES_CONF}"
echo "${VARIABLES_CONF}=\"${VARIABLES_CONF_EVAL}\""
done
;;

*)
ERROR "No conozco el ayudante '"${1}"', échale un ojo a la documentación (man canaima-desarrollador)"
;;
esac

exit 0
